### PR TITLE
Admin

### PR DESCRIPTION
**Pull request by lnmedrano**
## Summary

- Add endpoint POST /admin/users
- Add admin permissions for users
- Require admin-authentication for creating admin-users
- Tests for the endpoint

## Known Issues

- Repeated code in /app/models/user.js because linter does not support more than 4 parameters.

## Trello Card

https://trello.com/c/hPrZV8iz/17-usuario-administrador

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -4,20 +4,45 @@ const { validationError } = require('../errors'),
   { hashPassword, comparePasswords } = require('../helpers/hasher'),
   { createToken } = require('../helpers/token');
 
+const signUpWithAdminCondition = userData => {
```
Comment (SantiagoFoster) [inline app/controllers/users.js]: Remember that you should not have loose functions on the controller. A function like this would fit well inside an interactor. Read the rest of the comments first tho, you might end up not even needing this function.

File: app/controllers/users.js [hunk 2]
```diff
@@ -4,20 +4,45 @@ const { validationError } = require('../errors'),
...
+  const { firstName, lastName, email, password, isAdmin } = userData;
+  if (isAdmin) {
+    logger.info(`Creating user ${firstName} as admin`);
+    return hashPassword(password).then(hash => User.createUserAdmin(firstName, lastName, email, hash));
+  }
+  logger.info(`Creating user ${firstName}`);
+  return hashPassword(password).then(hash => User.createUser(firstName, lastName, email, hash));
```
Comment (SantiagoFoster) [inline app/controllers/users.js]: I like the intent here, but why not just have one `createUser` with an added `isAdmin` param? You can even set said param to a default, like `User.createUser = (firstName, lastName, email, hash, isAdmin = false) => ...`, so you don't have to change most of the code that already uses `createUser`.
(If you call a function with less params than it actually has, it automagically fills them with either `undefined` or a default if it exists)

File: app/controllers/users.js [hunk 3]
```diff
@@ -4,20 +4,45 @@ const { validationError } = require('../errors'),
...
+  signUpWithAdminCondition({ firstName, lastName, email, password, isAdmin: false })
     .then(([user, created]) => {
       if (created) {
         res.status(200).send(`User ${user.firstName} created`);
       } else {
         throw validationError('Email already used');
       }
     })
-    .catch(next);
+    .catch(error => next(error));
```
Comment (SantiagoFoster) [inline app/controllers/users.js]: Why this change? You are doing exactly the same thing with more code.

File: app/middlewares/validations/token.js [hunk 4]
```diff
@@ -1,22 +1,40 @@
...
+  } else {
+    next(validationError("User doesn't have needed permissions"));
+  }
+};
+
+exports.validateToken = (req, res, next) =>
+  getEmailFromToken(req.body.token)
+    .then(email => validateWithEmail(req.body.token, email))
+    .then(validated => resolveValidation(validated, next))
+    .catch(error => {
```
Comment (SantiagoFoster) [inline app/middlewares/validations/token.js]: You can remove the brackets `{}` here.
By the way, I like how you organized these functions!

File: app/models/user.js [hunk 5]
```diff
@@ -29,6 +29,10 @@ module.exports = (sequelize, DataTypes) => {
       password: {
         type: DataTypes.STRING,
         allowNull: false
+      },
+      isAdmin: {
+        type: DataTypes.BOOLEAN,
+        allowNull: true
```
Comment (SantiagoFoster) [inline app/models/user.js]: You could set a default to false, just in case

File: test/admin.test.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+            expect(foundUser.isAdmin).toBe(true);
+          });
+        }));
+  });
+
+  describe('Test create user admin', () => {
+    test('Create user admin with new email', () =>
+      hashPassword(correctPassword)
+        .then(pass =>
+          factory.create('userAdmin', {
```
Comment (SantiagoFoster) [inline test/admin.test.js]: you are creating this `userAdmin` for most of the tests. You should put it inside a `beforeEach()`to avoid repeating yourself so much. Same goes for any other commonly repeated test code.

File: test/admin.test.js [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        password: buildOptions.password,
+        isAdmin: false
+      }));
+
+      factory.define('userAdmin', User, buildOptions => ({
+        first_name: buildOptions.firstName,
+        last_name: buildOptions.lastName,
+        email: buildOptions.email,
+        password: buildOptions.password,
+        isAdmin: true
```
Comment (SantiagoFoster) [inline test/admin.test.js]: We usually have a separate file for defining factories. Its usually `test/support/factories.js`.
Also, you can just do 
```
      factory.define('userAdmin', User, buildOptions => ({
        first_name: chance.name(),
        last_name: chance.last(),
        email: chance.email({domain: "wolox.com"}),
        password: factory.chance('string', { pool: 'abcde' }),
        isAdmin: true
```
To randomly generate the params that are not integral to the factory (Its called 'userAdmin', so that`isAdmin: true` is the only critical thing). Take a look at the [chance](https://chancejs.com/index.html) library that factory-girl uses.

You can still overwrite params as needed in the same way you've been doing when you `.create`.
Keep this in mind to avoid unnecessary setup, like manually setting the first and last name. You are not going to check for them anyways.

File: test/admin.test.js [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+          })
+        )
+        .then(() =>
+          hashPassword(correctPassword).then(pass =>
+            factory.create('userAdmin', {
+              firstName,
+              lastName,
+              email: correctAdminEmail,
+              password: pass
+            })
```
Comment (SantiagoFoster) [inline test/admin.test.js]: Why create two userAdmins for this test?

File: app/controllers/users.js [hunk 9]
```diff
@@ -4,20 +4,45 @@ const { validationError } = require('../errors'),
...
+      if (created) {
+        res.status(200).send(`User ${user.firstName} created as admin`);
+      } else {
+        User.makeAdmin(email).then(() => {
+          logger.info(`User ${firstName} has admin permissions`);
+          res.status(200).send(`User ${firstName} has admin permissions`);
+        });
+      }
+    })
+    .catch(error => next(error));
```
Comment (SantiagoFoster) [inline app/controllers/users.js]: As I said in other comment, this could just be `.catch(next)`

**Review by SantiagoFoster — CHANGES_REQUESTED** [review]

File: app/controllers/users.js (hunk 2 above)
Comment (lnmedrano) [reply app/controllers/users.js]: That is what I tried to do first, but Linter threw this error:
`Arrow function has too many parameters (5). Maximum allowed is 4.eslint(max-params)`.
So the two options were receiving an object as parameter, or having that two functions. I will change it to receive an object, and having one function.

Comment (SantiagoFoster) [reply app/controllers/users.js]: Even if linter is crying about max params, you can send the rest of the params as an object for example. So the function could be `User.createUser = (userParams, isAdmin = false)`. However, in this case it would force you to change how the fucntion is used which is a bit annoying.
I'm not too keen on that max params rule myself. You can ignore this one for example by doing 

\/* eslint-disable max-params \*/
Rule violating code here
\/* eslint-enable max-params \*/

Be very careful with this however and only do it when you have a justifiable reason to do so.

File: app/models/user.js [hunk 10]
```diff
@@ -32,20 +32,21 @@ module.exports = (sequelize, DataTypes) => {
...
-        allowNull: true
+        allowNull: true,
+        defaultValue: false
       }
     },
     { underscored: true }
   );
 
-  User.createUser = (firstName, lastName, email, password) =>
+  User.createUser = ({ firstName, lastName, email, password, isAdmin = false }) =>
```
Comment (SantiagoFoster) [inline app/models/user.js:42]: 👍

File: package.json [hunk 11]
```diff
@@ -58,6 +58,7 @@
     "bcrypt": "^3.0.6",
     "bcryptjs": "^2.4.3",
     "body-parser": "^1.18.2",
+    "chance": "^1.0.18",
```
Comment (SantiagoFoster) [inline package.json]: If you are using chance directly from the `factory` then this is not needed, but uninstall it for next PR.

**Review by SantiagoFoster — APPROVED** [review]
Great!
